cartographer/mapping/internal/2d/normal_estimation_2d.cc

```diff
@@ -0,0 +1,121 @@
+template <typename T>
+T CircularMean(const std::vector<T>& angles) {
+  T summed_y = 0;
+  T summed_x = 0;
+  for (const T angle : angles) {
+    summed_y += std::sin(angle);
+    summed_x += std::cos(angle);
+  }
+  return ceres::atan2(summed_y, summed_x);
+}
+
+float Angle(const Eigen::Vector3f& v) { return std::atan2(v[1], v[0]); }
+
+// Estimate the normal of an 'observation' as the arithmetic mean of the the
+// normals of the vectors from 'observation' to each 'neighbor'.
+float EstimateNormal(const Eigen::Vector3f& observation,
+                     const std::vector<Eigen::Vector3f>& neighbors,
+                     const Eigen::Vector3f& observation_origin) {
+  if (neighbors.empty()) {
+    return Angle(observation_origin - observation);
+  }
+  std::vector<float> normals;
+  const Eigen::Vector3f origin_to_observation =
+      observation_origin - observation;
+  for (const auto& neighbor : neighbors) {
+    const Eigen::Vector3f tangent = observation - neighbor;
+    Eigen::Vector3f normal = {-tangent[1], tangent[0], tangent[2]};
+    // Ensure normal points towards 'observation_origin'.
+    if (normal.dot(origin_to_observation) < 0) {
+      normal = -normal;
+    }
+    float normal_angle = Angle(normal);
+    normals.push_back(normal_angle);
+  }
+  return CircularMean(normals);
+}
+}  // namespace
+
+proto::NormalEstimationOptions2D CreateNormalEstimationOptions2D(
+    common::LuaParameterDictionary* parameter_dictionary) {
+  proto::NormalEstimationOptions2D options;
+  options.set_num_normal_samples(
+      parameter_dictionary->GetInt("num_normal_samples"));
+  options.set_sample_radius(parameter_dictionary->GetDouble("sample_radius"));
+  CHECK_GT(options.num_normal_samples(), 0);
+  CHECK_GT(options.sample_radius(), 0.0);
+  return options;
+}
+
+// Estimates the normal for each 'return' in 'range_data'.
+// Assumes the angles in the range data returns are sorted with respect to
+// the orientation of the vector from 'origin' to 'return'.
+void EstimateNormals(
+    const sensor::RangeData& range_data, std::vector<float>* normals,
+    const proto::NormalEstimationOptions2D& normal_estimation_options) {
+  CHECK(normals != nullptr);
+  const size_t max_num_samples = normal_estimation_options.num_normal_samples();
+  const float sample_radius = normal_estimation_options.sample_radius();
+  for (size_t i_hit = 0; i_hit < range_data.returns.size(); ++i_hit) {
+    const Eigen::Vector3f& hit = range_data.returns[i_hit];
+    size_t neighbors_lower_bound = i_hit;
+    while (neighbors_lower_bound > 0 &&
+           i_hit - neighbors_lower_bound < max_num_samples / 2 &&
+           (hit - range_data.returns[neighbors_lower_bound - 1]).norm() <
+               sample_radius) {
+      neighbors_lower_bound--;
+    }
+    size_t neighbors_upper_bound = i_hit;
+    while (neighbors_upper_bound < range_data.returns.size() - 1 &&
+           neighbors_upper_bound - i_hit <
+               max_num_samples / 2 + max_num_samples % 2 &&
```

Comment:
rewrite this to (max_num_samples + 1) / 2 or similar to remove the modulo operation that is very hard to read